src/frontend/src/optimizer/plan_node/stream_sink.rs

```diff
@@ -112,35 +113,50 @@ impl StreamSink {
         definition: String,
         properties: WithOptions,
     ) -> Result<SinkDesc> {
+        const DOWNSTREAM_PK_KEY: &str = "primary_key";
+
         let distribution_key = input.distribution().dist_column_indices().to_vec();
         let sink_type = Self::derive_sink_type(input.append_only(), &properties)?;
-        let (pk, stream_key) = derive_pk(input, user_order_by, &columns);
-
-        if sink_type == SinkType::Upsert && pk.is_empty() {
-            return Err(ErrorCode::SinkError(Box::new(Error::new(
-                ErrorKind::InvalidInput,
-                "No primary key for the upsert sink. Please include the primary key explicitly in sink definition or make the sink append-only.",
-            )))
-            .into());
-        }
+        let (pk, _) = derive_pk(input, user_order_by, &columns);
```

Comment:
IIRC, no need to derive pk now? maybe mark the field deprecated.

Reply:
cc. @st1page

Reply:
I think so. But I prefer to store "down_stream_pk" and "plan_original_pk" in the SinkCatalog proto at least for debugging